rfcs/0017-tx-valid-since/0017-tx-valid-since.md

```diff
@@ -15,97 +15,246 @@ This RFC suggests adding a new consensus rule to prevent a cell to be spent befo
 
-The median block time calculated from the last 11 blocks timestamp (include tip block), we pick the older timestamp as median if blocks number is not enough and is odd, the details behavior defined as the following code:
+The median block time calculated from the past 11 blocks timestamp (from block's parent), we pick the older timestamp as median if blocks number is not enough and is odd, the details behavior defined as the following code:
 
 ``` rust
 pub trait BlockMedianTimeContext {
-    // number of last blocks, always return 11 in current consensus
-    fn block_count(&self) -> u32;
-    // Get block timestamp from block hash
-    fn timestamp(&self, hash: &H256) -> Option<u64>;
-    fn parent_hash(&self, hash: &H256) -> Option<H256>;
-    fn block_median_time(&self, hash: &H256) -> Option<u64> {
-        let count = self.block_count() as usize;
-        let mut block_times = Vec::with_capacity(count);
-        let mut current_hash = hash.to_owned();
-        for _ in 0..count {
-            match self.timestamp(&current_hash) {
-                Some(timestamp) => block_times.push(timestamp),
-                None => break,
-            }
-            match self.parent_hash(&current_hash) {
-                Some(hash) => current_hash = hash,
-                None => break,
-            }
-        }
-        block_times.sort_by(|a, b| b.cmp(a));
-        block_times.get(block_times.len() / 2).cloned()
+    fn median_block_count(&self) -> u64;
+    /// block timestamp
+    fn timestamp(&self, block_number: BlockNumber) -> Option<u64>;
+    /// ancestor timestamps from a block
+    fn ancestor_timestamps(&self, block_number: BlockNumber) -> Vec<u64> {
+        let count = self.median_block_count();
+        (block_number.saturating_sub(count)..=block_number)
+            .filter_map(|n| self.timestamp(n))
+            .collect()
+    }
+
+    /// get block median time
+    fn block_median_time(&self, block_number: BlockNumber) -> Option<u64> {
+        let mut timestamps: Vec<u64> = self.ancestor_timestamps(block_number);
+        timestamps.sort_by(|a, b| a.cmp(b));
+        // return greater one if count is even.
+        timestamps.get(timestamps.len() / 2).cloned()
     }
 }
 ```
 
 Validation of transaction `valid_since` defined as follow code:
 
 ``` rust
+const LOCK_TYPE_FLAG: u64 = 1 << 63;
+const TIME_TYPE_FLAG: u64 = 1 << 62;
+const TIMESTAMP_SCALAR: u64 = 9;
+const VALUE_MUSK: u64 = 0x00ff_ffff_ffff_ffff;
+
+#[derive(Copy, Clone, Debug)]
+struct ValidSince(u64);
+
+impl ValidSince {
+    pub fn is_absolute(self) -> bool {
+        self.0 & LOCK_TYPE_FLAG == 0
+    }
+
+    #[inline]
+    pub fn is_relative(self) -> bool {
+        !self.is_absolute()
+    }
+
+    fn metric_type_is_number(self) -> bool {
+        self.0 & TIME_TYPE_FLAG == 0
+    }
```

Comment:
Just `is_number` is enough, **metric_type** is implicit.

Reply:
The prefix is for distinguishing with `is_absolute` / `is_relative`.